test/functional/p2p_invalid_messages.py

```diff
@@ -42,13 +43,30 @@ def set_test_params(self):
         self.setup_clean_chain = True
 
     def run_test(self):
+        self.test_buffer()
         self.test_magic_bytes()
         self.test_checksum()
         self.test_size()
         self.test_msgtype()
         self.test_large_inv()
         self.test_resource_exhaustion()
 
+    def test_buffer(self):
+        self.log.info("Test message with header split across two buffers")
+        conn = self.nodes[0].add_p2p_connection(P2PDataStore())
+        # Create valid message
+        msg = conn.build_message(msg_ping(nonce=12345))
+        cut_pos = 12    # Chosen at an arbitrary position within the header
+        # Send message in two pieces
+        before = self.nodes[0].getnettotals()['totalbytesrecv']
+        conn.send_raw_message(msg[:cut_pos])
+        middle = self.nodes[0].getnettotals()['totalbytesrecv']
+        # If this assert fails, we've hit an unlikely race
+        # where the test framework sent a message in between the two halves
+        assert middle == before + cut_pos
+        conn.send_raw_message(msg[cut_pos:])
+        conn.sync_with_ping(timeout=1)
```

Comment:
I suggest you disconnect the connection at the end of this subtest, so that the node and test fixture are the same state at the end of each subtest as at the start.

Ideally, it should be possible to permute subtests in any order and have the test pass. That prevents brittleness like the asserts on peer numbers that you've had to change in an unrelated subtest here.